clients/src/test/java/org/apache/kafka/common/network/EchoServer.java

```diff
@@ -73,7 +74,9 @@ public void run() {
         try {
             while (true) {
                 final Socket socket = serverSocket.accept();
-                sockets.add(socket);
+                synchronized (sockets) {
+                    if (!closing) sockets.add(socket);
```

Comment:
Hi Ted, I quickly had a look at this change if you don't mind. My concern is the following:
Let's suppose that if right after `serverSocket.accept()` a `closeConnection()` gets scheduled and completes, then this block would indeed run without adding the current socket to a list but it would still start the thread with a leaked socket.
I think we shouldn't even accept any sockets or start threads if the server is being shut down.
Let me know if you disagree or I missed something :).